core/test/Cardano/Address/ScriptSpec.hs

```diff
@@ -447,14 +447,59 @@ spec = do
             validateScriptTemplate RecommendedValidation scriptTemplate `shouldBe` Left (WrongScript $ NotRecommended RedundantTimelocks)
 
         it "valid timelocks" $ do
-            let scriptTemplate = ScriptTemplate cosigners' (RequireAllOf [cosigner0, cosigner1, cosigner2 ,cosigner3, ActiveFromSlot 11, ActiveUntilSlot 15])
+            let scriptTemplate = ScriptTemplate cosigners' (RequireAllOf [cosigner0, cosigner1, cosigner2 ,cosigner3, ActiveFromSlot 11, ActiveUntilSlot 11])
             validateScriptTemplate RecommendedValidation scriptTemplate `shouldBe` Right ()
 
-        it "invalid timelocks - contradictory 1" $ do
+        it "valid timelocks when using all 1" $ do
             let scriptTemplate = ScriptTemplate cosigners' (RequireSomeOf 1
                     [ cosigner1
                     , RequireAnyOf [ cosigner2
-                                   , RequireAllOf [cosigner0, cosigner1, cosigner2 ,cosigner3, ActiveFromSlot 21, ActiveUntilSlot 20]
+                                   , RequireAllOf [cosigner0, cosigner1, cosigner2 ,cosigner3, ActiveFromSlot 21, ActiveUntilSlot 25]
+                                   ]
+                    ])
+            validateScriptTemplate RecommendedValidation scriptTemplate `shouldBe` Right ()
+
+        it "valid timelocks when using all 2" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireSomeOf 1
+                    [ cosigner1
+                    , RequireAnyOf [ cosigner2
+                                   , RequireAllOf [cosigner0, cosigner1, cosigner2 ,cosigner3, ActiveFromSlot 21, ActiveUntilSlot 21]
+                                   ]
+                    ])
+            validateScriptTemplate RecommendedValidation scriptTemplate `shouldBe` Right ()
+
+        it "invalid timelocks when using all" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireSomeOf 1
+                    [ cosigner1
+                    , RequireAnyOf [ cosigner2
+                                   , RequireAllOf [cosigner0, cosigner1, cosigner2 ,cosigner3, ActiveFromSlot 25, ActiveUntilSlot 21]
+                                   ]
+                    ])
+            validateScriptTemplate RecommendedValidation scriptTemplate `shouldBe` Left (WrongScript $ NotRecommended TimelockTrap)
+
+        it "valid timelocks when using any" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireSomeOf 1
+                    [ cosigner1
+                    , RequireAnyOf [ cosigner2
+                                   , RequireAnyOf [cosigner0, cosigner1, cosigner2 ,cosigner3, ActiveFromSlot 25, ActiveUntilSlot 21]
```

Comment:
I don't get why this is considered valid ?

Reply:
well, to validate this part the following is needed:
(a) any time one of signatures of any cosigner : 0, 1, 2, 3

Reply:
Okay. My reasoning was that basically having scripts in the shape of `any [signature, timelock]` is generally unsound. When specifying a timelock, I think you almost always want to use `all` or `at_least n` with n >= 2.